### PR TITLE
Bugfix: let source maintainer revoke a request

### PR DESCRIPTION
**Pull request by ncounter**
Fixes https://github.com/openSUSE/open-build-service/issues/16105

### CODE REVIEW

File: src/api/app/policies/bs_request_policy.rb [hunk 1]
```diff
@@ -20,7 +20,7 @@ def add_reviews?
   end
 
   def revoke_request?
-    author? && record.state.in?(%i[new review declined])
+    (author? || record.is_target_maintainer?(user)) && record.state.in?(%i[new review declined])
```
Comment (hennevogel) [inline src/api/app/policies/bs_request_policy.rb:23]:
```suggestion
    return false if [:new, :review, :declined].exclude?(record.state)

    author? || record.is_target_maintainer?(user)
```

**Review by hennevogel — CHANGES_REQUESTED** [review]

File: src/api/app/policies/bs_request_policy.rb (hunk 1 above)
Comment (hennevogel) [reply src/api/app/policies/bs_request_policy.rb:23]: What does `BsRequestPermissionCheck` say about this case?

**Comment by adrianschroeter** [comment]
this might be a not wanted regression, please see the reported issue.

**Comment by ncounter** [comment]
Based on @adrianschroeter and @hennevogel comments, I moved the implementation from _allowing users with write permission in the target_ to _allowing users with write permission in the sources_ instead, according to already existing rules.

**Review by hennevogel — CHANGES_REQUESTED** [review]

File: src/api/app/policies/bs_request_policy.rb [hunk 2]
```diff
@@ -9,8 +9,9 @@ def create?
   end
 
   def handle_request?
-    is_target_maintainer = record.is_target_maintainer?(user)
-    record.state.in?(%i[new review declined]) && (is_target_maintainer || author?)
+    return false if %i[new review declined].exclude?(record.state)
+
+    author? || record.is_target_maintainer?(user) || record.is_source_maintainer?(user)
```
[the remark this answers is not in the document]
Comment (ncounter) [reply src/api/app/policies/bs_request_policy.rb:14]: It allows the `RequestDecisionComponent` to render first (https://github.com/openSUSE/open-build-service/blob/master/src/api/app/components/request_decision_component.rb#L13). Later on inside the component every request action is warded by an additional check, like the `revoke_request?` one we need here.

[the remark this answers is not in the document]
Comment (hennevogel) [reply src/api/app/policies/bs_request_policy.rb:14]: I don't like that we deviate from the non beta UI but okay...

**Review by hennevogel — APPROVED** [review]

**Pull request merged**